src/fdTransfer.cpp

```diff
@@ -0,0 +1,343 @@
+};
+struct fd_request {
+    enum request_type type;
+    unsigned int length;
+#define MAX_REQUEST_LENGTH 2048
+    unsigned int request_id;
+};
+
+struct perf_fd_request {
+    struct fd_request header;
+    pid_t tid;
+    struct perf_event_attr attr;
+};
+
+#define TMP_KALLSYMS_PATH "/tmp/async-profiler-kallsyms"
+
+int FdTransfer::_listener = -1;
+int FdTransfer::_peer = -1;
+unsigned int FdTransfer::_request_id = 0;
+
+bool FdTransfer::serveRequests() {
+    while (1) {
+        unsigned char request_buf[MAX_REQUEST_LENGTH];
+        struct fd_request *header = (struct fd_request *)request_buf;
+
+        ssize_t ret = recv(_peer, header, sizeof(*header), 0);
+        if (ret < 0) {
+            perror("recv()");
+            break;
+        } else if (ret == 0) {
+            // EOF means done
+            return true;
+        } else if (ret != sizeof(*header)) {
+            // TODO truncated reads
+            fprintf(stderr, "truncated request header\n");
+            break;
+        }
+
+        if (header->length > MAX_REQUEST_LENGTH) {
+            fprintf(stderr, "too large request: %zd\n", ret);
+            break;
+        } else if (header->length < sizeof(*header)) {
+            fprintf(stderr, "bogus header length\n");
+            break;
+        }
+
+        const size_t left = header->length - sizeof(*header);
+        if (left > 0) {
+            ret = recv(_peer, &header[1], header->length, 0);
+            if (ret < 0) {
+                perror("recv");
+                break;
+            } else if (ret != left) {
+                // TODO truncated reads
+                fprintf(stderr, "truncated request\n");
+                break;
+            }
+        }
+
+        switch (header->type) {
+        case PERF_FD: {
+            // TODO validate 'tid' is indeed a thread of 'nspid'
+            struct perf_fd_request *request = (struct perf_fd_request*)header;
+
+            int perf_fd = syscall(__NR_perf_event_open, &request->attr, request->tid, -1, -1, 0);
+            if (perf_fd == -1) {
+                perror("perf_event_open()");
+            }
+
+            sendFd(perf_fd, request->header.request_id);
+            break;
+        }
+
+        case KALLSYMS_FD: {
+            // can't directly pass the fd of /proc/kallsyms, because before Linux 4.15 the permission check
+            // was conducted on each read.
+            // it's simpler to copy the file to a temporary location and pass the fd of it (compared to passing the
+            // entire contents over the peer socket)
+            {
+                // TODO from Linux 3.15, we can use memfds here
+                std::ifstream src("/proc/kallsyms", std::ios::binary);
+                // TODO this creates the file with 0611, it should be 0600.
+                std::ofstream dst(TMP_KALLSYMS_PATH, std::ios::binary);
+                dst << src.rdbuf();
+            }
+
+            int kallsyms_fd = open(TMP_KALLSYMS_PATH, O_RDONLY);
```

Comment:
In case 2 async-profilers run simultaneously, this part is racy - one may override the temp file of another. The options I see:
1. Include target PID in the filename - this avoids the chance of collisions, but creates more temporary files. Since we don't cache them - we can just remove them once `fdtransfer` goes down.
2. Use a `pipe(2)` - I didn't like this approach at first, because `fdtransfer` is synchronous and pipe buffers are, in general, small and will not fit the whole `/proc/kallsyms`. If `fdtransfer` is bound to run as root - I'm pretty sure root has no limits of pipe buffer sizes? Need to check that.
3. We can give up on passing the fd and just pass the filename. But we still have this "race" - we need to copy the file to somewhere accessible by async-profiler, and it has to be unique in case there are numerous async-profilers.

I'll check the idea pipe first.

Reply:
Easiest done with including the target PID. I don't like the synchronous pipes thing (which has hidden assumptions, like the one that async-profiler should consume the entire file immediately upon requesting it...)

Reply:
Done with the files

Reply:
Also, now I move into the mount NS; without doing that, this was flawed - multiple `fdtransfer` programs running in different PID NSs could have the same NSpid so the resulting path is the same; but they are also in the same mount NS....